vulnerabilities/package_managers_2.py

```diff
@@ -0,0 +1,377 @@
+        for child in xml_response.getroot().iter():
+            if child.tag == "version" and child.text:
+                all_versions.add(LegacyVersion(child.text))
+
+        return all_versions
+
+
+class NugetVersionAPI(VersionAPI):
+
+    package_type = "nuget"
+
+    def fetch(self, pkg) -> None:
+        endpoint = self.nuget_url(pkg)
+        resp = get_response(url=endpoint)
+        if not resp:
+            self.cache[pkg] = set()
+            return
+        self.cache[pkg] = self.extract_versions(resp)
+
+    @staticmethod
+    def nuget_url(pkg_name: str) -> str:
+        pkg_name = pkg_name.lower().strip()
+        base_url = "https://api.nuget.org/v3/registration5-semver1/{}/index.json"
+        return base_url.format(pkg_name)
+
+    @staticmethod
+    def extract_versions(resp: dict) -> Set[LegacyVersion]:
+        all_versions = set()
+        try:
+            for entry_group in resp["items"]:
+                for entry in entry_group["items"]:
+                    all_versions.add(
+                        LegacyVersion(
+                            value=entry["catalogEntry"]["version"],
+                            release_date=dateparser.parse(entry["catalogEntry"]["published"]),
+                        )
+                    )
+        # FIXME: json response for YamlDotNet.Signed triggers this exception.
+        # Some packages with many versions give a response of a list of endpoints.
+        # In such cases rather, we should collect data from those endpoints.
+        except KeyError:
+            pass
+
+        return all_versions
+
+
+class GoproxyVersionAPI(VersionAPI):
+
+    package_type = "golang"
+
+    def __init__(self, cache: MutableMapping[str, Set[LegacyVersion]] = None):
+        super().__init__(cache)
+        self.module_name_by_package_name = {}
+
+    @staticmethod
+    def trim_url_path(url_path: str) -> Optional[str]:
+        """
+        Return a trimmed Go `url_path` removing trailing
+        package references and keeping only the module
+        references.
+
+        Github advisories for Go are using package names
+        such as "https://github.com/nats-io/nats-server/v2/server"
+        (e.g., https://github.com/advisories/GHSA-jp4j-47f9-2vc3 ),
+        yet goproxy works with module names instead such as
+        "https://github.com/nats-io/nats-server" (see for details
+        https://golang.org/ref/mod#goproxy-protocol ).
+        This functions trims the trailing part(s) of a package URL
+        and returns the remaining the module name.
+        For example:
+        >>> module = "https://github.com/xx/a"
+        >>> assert GoproxyVersionAPI.trim_url_path("https://github.com/xx/a/b") == module
+        """
+        # some advisories contains this prefix in package name, e.g. https://github.com/advisories/GHSA-7h6j-2268-fhcm
+        if url_path.startswith("https://pkg.go.dev/"):
+            url_path = url_path.removeprefix("https://pkg.go.dev/")
```

Comment:
This works only on Python 3.9 ... this is a bug at multiple levels:
1. we do not want to support exclusively the 3.9 syntax now, since our minimum Python version is 3.8
2. this is clear that this is not tested at all... we need a test for this
3. we may not have a CI test running on Python 3.8

There are some possible bugs:
```
>>> trim_url_path("https://pkg.go.dev/github.com/containous/traefik/v2")
'github.com/containous/traefik'
>>> trim_url_path("https://pkg.go.dev/github.com/containous")
'github.com'
>>> trim_url_path("https://pkg.go.dev/github.com")
```

I reckon this is a carry over from the old package_managers, but this whole thing is both undertested and may not make sense. We need tests :)